packages/@lwc/template-compiler/src/shared/ir.ts

```diff
@@ -42,105 +141,293 @@ export function createElement(original: parse5.Element): IRElement {
         throw new Error('Invalid element AST node. Missing source code location.');
     }
 
+    return location;
+}
+
+export function parseSourceLocation(location: parse5.Location): SourceLocation {
     return {
-        type: 'element',
-        tag: original.tagName,
-        namespace: original.namespaceURI,
+        startLine: location.startLine,
+        startColumn: location.startCol,
+        endLine: location.endLine,
+        endColumn: location.endCol,
+        start: location.startOffset,
+        end: location.endOffset,
+    };
+}
+
+export function literal(value: string | boolean): Literal {
+    return {
+        type: LWCNodeType.Literal,
+        value,
+    };
+}
+
+export function forEach(
+    name: string,
+    expression: Expression,
+    location: SourceLocation,
+    item: Identifier,
+    index?: Identifier
+): ForEach {
+    return {
+        type: LWCNodeType.ForEach,
+        name,
+        expression,
+        location,
         children: [],
+        item,
+        index,
+    };
+}
+
+export function forOf(
+    name: string,
+    expression: Expression,
+    iterator: Identifier,
+    location: SourceLocation
+): ForOf {
+    return {
+        type: LWCNodeType.ForOf,
+        name,
+        expression,
+        iterator,
         location,
-        __original: original,
+        children: [],
     };
 }
 
-export function createText(original: parse5.TextNode, value: string | TemplateExpression): IRText {
-    if (!original.sourceCodeLocation) {
-        throw new Error('Invalid text AST node. Missing source code location.');
-    }
+export function ifBlock(
+    name: string,
+    location: SourceLocation,
+    modifier: string,
+    condition: Expression
+): IfBlock {
+    return {
+        type: LWCNodeType.IfBlock,
+        name,
+        location,
+        children: [],
+        modifier,
+        condition,
+    };
+}
 
+export function eventListener(
+    name: string,
+    handler: Expression,
+    location: SourceLocation
+): EventListener {
     return {
-        type: 'text',
+        type: LWCNodeType.EventListener,
+        name,
+        handler,
+        location,
+    };
+}
+
+export function keyDirective(value: Expression, location: SourceLocation): KeyDirective {
+    return {
+        type: LWCNodeType.Directive,
+        name: LWCNodeType.Key,
         value,
-        location: original.sourceCodeLocation,
+        location,
     };
 }
 
-export function createComment(original: parse5.CommentNode, value: string): IRComment {
-    if (!original.sourceCodeLocation) {
-        throw new Error('Invalid comment AST node. Missing source code location.');
-    }
+export function dynamicDirective(value: Expression, location: SourceLocation): DynamicDirective {
+    return {
+        name: LWCNodeType.Dynamic,
+        value,
+        type: LWCNodeType.Directive,
```

Comment:
To remain consistent with the other helper methods, I would move all the type fields as the first entry in the object literal.